okhttp/src/test/java/io/grpc/okhttp/AsyncFrameWriterTest.java

```diff
@@ -0,0 +1,137 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AsyncFrameWriterTest {
+
+  @Mock private Socket socket;
+  @Mock private FrameWriter frameWriter;
+
+  private QueueingExecutor queueingExecutor = new QueueingExecutor();
+  private TransportExceptionHandler transportExceptionHandler =
+      new EscalatingTransportErrorHandler();
+
+  private AsyncFrameWriter asyncFrameWriter;
+
+  @Before
+  public void setUp() throws Exception {
+    asyncFrameWriter =
+        new AsyncFrameWriter(transportExceptionHandler, new SerializingExecutor(queueingExecutor));
+    asyncFrameWriter.becomeConnected(frameWriter, socket);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    queueingExecutor.clear();
+  }
+
+  @Test
+  public void noCoalesceRequired() throws IOException {
+    asyncFrameWriter.ping(true, 0, 1);
+    asyncFrameWriter.flush();
+    queueingExecutor.runAll();
+
+    verify(frameWriter, times(1)).ping(anyBoolean(), anyInt(), anyInt());
+    verify(frameWriter, times(1)).flush();
+  }
+
+  @Test
+  public void flushCoalescing_shouldNotMergeTwoDistinctFlushes() throws IOException {
+    asyncFrameWriter.ping(true, 0, 1);
+    asyncFrameWriter.flush();
+    queueingExecutor.runAll();
+
+    asyncFrameWriter.ping(true, 0, 2);
+    asyncFrameWriter.flush();
+    queueingExecutor.runAll();
+
+    verify(frameWriter, times(2)).ping(anyBoolean(), anyInt(), anyInt());
+    verify(frameWriter, times(2)).flush();
+  }
+
+  @Test
+  public void flushCoalescing_shouldMergeTwoQueuedFlushes() throws IOException {
+    asyncFrameWriter.ping(true, 0, 1);
+    asyncFrameWriter.flush();
+    asyncFrameWriter.ping(true, 0, 2);
+    asyncFrameWriter.flush();
+
+    queueingExecutor.runAll();
+
+    InOrder inOrder = inOrder(frameWriter);
+    inOrder.verify(frameWriter, times(2)).ping(anyBoolean(), anyInt(), anyInt());
+    inOrder.verify(frameWriter).flush();
+  }
+
+  /**
+   * Executor queues incoming runnables instead of running it. Runnables can be invoked via {@link
+   * QueueingExecutor#runAll} in serial order.
+   */
+  private static class QueueingExecutor implements Executor {
+
+    private final Queue<Runnable> runnables = new ConcurrentLinkedQueue<Runnable>();
+
+    @Override
+    public void execute(Runnable command) {
+      runnables.add(command);
+    }
+
+    public void runAll() {
+      Runnable r;
+      while ((r = runnables.poll()) != null) {
+        r.run();
+      }
+    }
+
+    public void clear() {
+      runnables.clear();
+    }
+  }
+
+  /** Rethrows as Internal error. */
+  private static class EscalatingTransportErrorHandler implements TransportExceptionHandler {
+
+    @Override
+    public void onException(Throwable throwable) {
+      throw Status.INTERNAL.asRuntimeException();
```

Comment:
This throws away the Throwable, and we don't use StatusRuntimeException too much like this. How about just `new AssertionError(throwable)`?